distutils/extension.py

```diff
@@ -20,138 +22,161 @@
+    list of command-line arguments, but for other platforms it could
+    be anything.
     """
 
-    # When adding arguments to this constructor, be sure to update
-    # setup_keywords in core.py.
-    def __init__(
-        self,
-        name: str,
-        sources: Iterable[str | os.PathLike[str]],
-        include_dirs: list[str] | None = None,
-        define_macros: list[tuple[str, str | None]] | None = None,
-        undef_macros: list[str] | None = None,
-        library_dirs: list[str] | None = None,
-        libraries: list[str] | None = None,
-        runtime_library_dirs: list[str] | None = None,
-        extra_objects: list[str] | None = None,
-        extra_compile_args: list[str] | None = None,
-        extra_link_args: list[str] | None = None,
-        export_symbols: list[str] | None = None,
-        swig_opts: list[str] | None = None,
-        depends: list[str] | None = None,
-        language: str | None = None,
-        optional: bool | None = None,
-        **kw,  # To catch unknown keywords
-    ):
-        if not isinstance(name, str):
+    extra_link_args: list[str] = field(default_factory=list)
+    """
+    any extra platform- and compiler-specific information to use
+    when linking object files together to create the extension (or
+    to create a new static Python interpreter).  Similar
+    interpretation as for 'extra_compile_args'.
+    """
+
+    export_symbols: list[str] = field(default_factory=list)
+    """
+    list of symbols to be exported from a shared extension.  Not
+    used on all platforms, and not generally necessary for Python
+    extensions, which typically export exactly one symbol: "init" +
+    extension_name.
+    """
+
+    swig_opts: list[str] = field(default_factory=list)
+    """
+    any extra options to pass to SWIG if a source file has the .i
+    extension.
+    """
+
+    depends: list[str] = field(default_factory=list)
+    """list of files that the extension depends on"""
+
+    language: str | None = None
+    """
+    extension language (i.e. "c", "c++", "objc"). Will be detected
+    from the source extensions if not provided.
+    """
+
+    optional: bool = False
+    """
+    specifies that a build failure in the extension should not abort the
+    build process, but simply not install the failing extension.
+    """
+
+
+# Legal keyword arguments for the Extension constructor
+_safe = tuple(f.name for f in fields(_Extension))
+
+
+@dataclass(init=True if TYPE_CHECKING else False)  # type: ignore[literal-required]
+class Extension(_Extension):
+    if not TYPE_CHECKING:
+
+        def __init__(self, *args, **kwargs):
+            extra = {repr(k): kwargs.pop(k) for k in tuple(kwargs) if k not in _safe}
+            if extra:
+                msg = f"""
+                Please remove unknown `Extension` options: {','.join(extra)}
+                this kind of usage is deprecated and may cause errors in the future.
+                """
+                warnings.warn(msg)
```

Comment:
If/when lenient behaviour is not needed the whole custom `__init__` can be removed (and the classes `_Extension/Extension` can be unified.